File: 068/main.js
/**
 * @param {string[]} words
 * @param {number} maxWidth
 * @return {string[]}
 */
var fullJustify = function (words, maxWidth) {
  const len = words.length
  let res = []
  let arr = []
  let sum = words[0].length
  let mid = ''
  if (!sum) {
    while (maxWidth--) mid += ' '
    res.push(mid)
  } else {
    arr.push(words[0])
  }
  for (let i = 1; i < len; i++) {
    if (sum + words[i].length + 1 <= maxWidth) {
      sum += words[i].length + 1
      arr.push(' ' + words[i])
    } else {
      let space = maxWidth - sum
      mid = ''
      if (arr.length === 1) {
        mid += arr[0]
        while (space--) mid += ' '
        res.push(mid)
      } else {
        let num = Math.floor(space / (arr.length - 1))
        let point = space % (arr.length - 1)
        for (let j = 0; j < arr.length - 1; j++) {
          mid += arr[j]
          for (let k = 0; k < num; k++) {
            mid += ' '
          }
          if (j < point) mid += ' '
        }
        mid += arr[arr.length - 1]
        res.push(mid)
      }
      sum = words[i].length
      arr.splice(0, arr.length)
      arr.push(words[i])
    }
  }
  if (sum) {
    mid = ''
    for (let i = 0; i < arr.length; i++) {
      mid += arr[i]
    }
    let num = maxWidth - mid.length
    while (num--) mid += ' '
    res.push(mid)
  }
  return res
}

console.log(fullJustify(['This', 'is', 'an', 'example', 'of', 'text', 'justification.'], 16))
console.log(fullJustify(['a'], 1))
console.log(fullJustify(['What', 'must', 'be', 'shall', 'be.'], 12))
console.log(fullJustify([''], 12))
console.log(fullJustify([''], 0))
console.log(fullJustify(['a', 'b', 'c', 'd', 'e'], 1))
console.log(fullJustify(['Listen', 'to', 'many,', 'speak', 'to', 'a', 'few.'], 6))

File: 068/main.cpp
#include <iostream>
#include <vector>
using namespace std;

class Solution
{
public:
  vector<string> fullJustify(vector<string> &words, int maxWidth)
  {
    int len = words.size();
    vector<string> res;
    vector<string> arr;
    int sum = words[0].size();
    string mid = "";
    if (!sum)
    {
      while (maxWidth--)
        mid += ' ';
      res.push_back(mid);
    }
    else
    {
      arr.push_back(words[0]);
    }
    for (int i = 1; i < len; i++)
    {
      if (sum + words[i].size() + 1 <= maxWidth)
      {
        sum += words[i].size() + 1;
        mid = ' ' + words[i];
        arr.push_back(mid);
      }
      else
      {
        int space = maxWidth - sum;
        mid = "";
        if (arr.size() == 1)
        {
          mid += arr[0];
          while (space--)
            mid += ' ';
          res.push_back(mid);
        }
        else
        {
          int num = space / (arr.size() - 1);
          int point = space % (arr.size() - 1);
          for (int j = 0; j < arr.size() - 1; j++)
          {
            mid += arr[j];
            for (int k = 0; k < num; k++)
            {
              mid += ' ';
            }
            if (j < point)
              mid += ' ';
          }
          mid += arr[arr.size() - 1];
          res.push_back(mid);
        }
        sum = words[i].size();
        arr.clear();
        arr.push_back(words[i]);
      }
    }
    if (sum)
    {
      mid = "";
      for (int i = 0; i < arr.size(); i++)
      {
        mid += arr[i];
      }
      int num = maxWidth - mid.size();
      while (num--)
        mid += ' ';
      res.push_back(mid);
    }
    return res;
  }
};

int main()
{
  Solution sol;
  vector<string> words, res;
  words.push_back("Listen");
  words.push_back("to");
  words.push_back("many");
  words.push_back("speak");
  words.push_back("to");
  words.push_back("a");
  words.push_back("few.");
  res = sol.fullJustify(words, 6);
  for (int i = 0; i < res.size(); i++)
  {
    cout << res[i] << endl;
  }
  return 0;
}
